crates/edr_provider/src/data.rs

```diff
@@ -1922,6 +1914,141 @@ impl<LoggerErrorT: Debug, TimerT: Clone + TimeSinceEpoch> ProviderData<LoggerErr
         Ok(transaction_hash)
     }
 
+    /// Returns the block header and the transactions from the block that
+    /// contains the transaction for debug trace transactions.
+    fn block_data_for_debug_trace_transaction(
+        &self,
+        transaction_hash: &B256,
+    ) -> Result<(Header, Vec<Signed>), ProviderError<LoggerErrorT>> {
+        // This is a hack to make `debug_traceTransaction` return a helpful error
+        // message in fork mode if there is a transaction in the block whose
+        // type is not supported or skip that transaction if an environment
+        // variable is set. This hack is only necessary until proper multichain
+        // support. https://github.com/NomicFoundation/edr/issues/570
+        self.rpc_client
+            .as_ref()
+            .and_then(|rpc_client| {
+                // Use `Result<Option>` in `block_in_place` to be able to short-circuit with
+                // `?`, but we want `Option<Result>` in the end as we're flat
+                // mapping an `Option`.
+                tokio::task::block_in_place::<_, Result<Option<_>, _>>(|| {
+                    // If the transaction is not found in the remote by the provided hash, there are
+                    // two possibilities:
+                    // 1. The transaction is from a local block. In this case, it must have valid
+                    //    transaction types, so we have nothing to do.
+                    // 2. There is no transaction with the provided hash. In this case the abstract
+                    //    interface will return an error.
+                    self.runtime_handle
+                        .block_on(rpc_client.get_transaction_by_hash(*transaction_hash))?
+                        .and_then::<Result<_, _>, _>(|transaction| {
+                            transaction
+                                .block_hash
+                                .ok_or_else(|| {
+                                    // If the transaction doesn't have a block hash, we treat the
+                                    // transaction hash as invalid.
+                                    ProviderError::<LoggerErrorT>::InvalidTransactionHash(
```

Comment:
In `debug_trace_transaction`, we return `DebugTraceError::InvalidTransactionHash`. To retain the same behaviour, this should probably match that error.

The same applies for the occurrences below

Reply:
We returned `ProviderError::<LoggerErrorT>::InvalidTransactionHash` previously from `ProviderData::debug_trace_transaction` when a transaction was not found: https://github.com/NomicFoundation/edr/pull/606/files#diff-03b2ca6fd82ba7931a70b3b1ef672aa25b18c86bf643d21a930efa2691e701cfL618-L621

`DebugTraceError::InvalidTransactionHash` is only returned from `edr_evm::debug_trace_transaction` if the arguments passed to it from `ProviderData::debug_trace_transaction` are wrong.

So `DebugTraceError::InvalidTransactionHash` is returned in case of a bug in our codebase, while `ProviderError::<LoggerErrorT>::InvalidTransactionHash` is returned if the user supplies wrong data. This PR retains that behavior, so I think it's ok as is.

Reply:
Thank you for explaining